conans/model/graph_lock.py

```diff
@@ -0,0 +1,278 @@
+import json
+import os
+
+from conans.client.graph.graph import RECIPE_VIRTUAL, RECIPE_CONSUMER,\
+    BINARY_BUILD
+from conans.client.profile_loader import _load_profile
+from conans.errors import ConanException
+from conans.model.ref import PackageReference, ConanFileReference
+from conans.util.files import load, save
+from conans.model.options import OptionsValues
+from platform import node
+
+
+LOCKFILE = "conan.lock"
+
+
+class GraphLockFile(object):
+
+    def __init__(self, profile, graph_lock):
+        self.profile = profile
+        self.graph_lock = graph_lock
+
+    @staticmethod
+    def load(path):
+        if not path:
+            raise IOError("Invalid path")
+        p = os.path.join(path, LOCKFILE)
+        if not os.path.isfile(p):
+            raise ConanException("Missing conan.lock file: %s" % p)
+        content = load(p)
+        try:
+            return GraphLockFile.loads(content)
+        except Exception as e:
+            raise ConanException("Error parsing lockfile '{}': {}".format(p, e))
+
+    @staticmethod
+    def loads(text):
+        graph_json = json.loads(text)
+        profile = graph_json["profile"]
+        # FIXME: Reading private very ugly
+        profile, _ = _load_profile(profile, None, None)
+        graph_lock = GraphLock.from_dict(graph_json["graph_lock"])
+        graph_lock_file = GraphLockFile(profile, graph_lock)
+        return graph_lock_file
+
+    def save(self, folder):
+        filename = LOCKFILE
+        p = os.path.join(folder, filename)
+        serialized_graph_str = self.dumps()
+        save(p, serialized_graph_str)
+
+    def dumps(self):
+        result = {"profile": self.profile.dumps(),
+                  "graph_lock": self.graph_lock.as_dict()}
+        return json.dumps(result, indent=True)
+
+
+class GraphLockNode(object):
+    def __init__(self, pref, python_requires, options, modified):
+        self.pref = pref
+        self.python_requires = python_requires
+        self.options = options
+        self.modified = modified
+
+    @staticmethod
+    def from_dict(node_json):
+        json_pref = node_json["pref"]
+        pref = PackageReference.loads(json_pref) if json_pref else None
+        python_requires = node_json.get("python_requires")
+        if python_requires:
+            python_requires = [ConanFileReference.loads(ref) for ref in python_requires]
+        options = OptionsValues.loads(node_json["options"])
+        modified = node_json.get("modified")
+        return GraphLockNode(pref, python_requires, options, modified)
+
+    def as_dict(self):
+        result = {}
+        result["pref"] = self.pref.full_repr() if self.pref else None
+        result["options"] = self.options.dumps()
+        if self.python_requires:
+            result["python_requires"] = [r.full_repr() for r in self.python_requires]
+        if self.modified:
+            result["modified"] = True
+        return result
+
+
+class GraphLock(object):
```

Comment:
Need to know the motivation for these classes.

I agree that we need to store the _graph lock_ and have a memory representation of it, but here there are more things: we are at least duplicating business logic that was already available. If the plan is to substitute existing implementation, then I think it is worth considering a more standard representation of the graph so we can implement different algorithms on top of it.

Reply:
Need to keep things as much decoupled as possible. Premature engineering is evil, much better to have some duplication until behavior is much more stabilized, than refactoring and abstracting away too early.
This is not substituting anything, it is totally a new feature, not replacing or changing any existing functionality.

Reply:
So, the plan is to substitute the existing graph? Without entering into the details (and without having a look at how it is used) I see that both graphs have _neighbours_, _closure_, _inverse levels_,... and I'm assuming that they should have the same information, right? So, we are going to maintain both implementations during some releases.

It's ok, I just want to know. If this is going to be merged soon, I will need to add all the changes related to cross-building to this new graph too.

Reply:
No, this doesn't plan to substitute the existing dependency graph at all.

Yes, some concepts are very similar, indeed. Still the implementation is very different in the details and interfaces, so that is why better keep them decoupled.

Regarding cross-building, this should be totally transparent to it. This graph does not manage conanfiles or settings at all. It doesn't even model build_requires!

It is just some data that the other graph will use. I'd say this doesn't need any modification for the cross-building model.